React Revamp: Fetching Data in React Lesson (#25811)
TheOdinProject/curriculum: react/the_react_ecosystem/fetching-data.md

  );
~~~

Notice how Bio is taking an extra second to display? Their fetch requests should both take 1000ms to resolve so what's going on?? In React none of the hooks, JSX, or code inside of a component will be run until that component is due to render. Bio has to wait for the request inside of Profile to resolve before it starts rendering which means the request inside Bio isn't sent.

If we remove the short circuiting conditional that waits for imageURL Bio would send a request immediately, but that would mean abandoning our loading screen. Instead of compromising on design we can lift the request up the component tree and pass it's response as a prop to Bio.  

To see this in action go back to that embedded codesandbox and comment out the current `Profile` and `Bio` components and uncomment the currently commented ones. 

Now we have both requests firing as soon as Profile renders. The request for imageURL resolves 2 seconds before the bioText request and our div containing <Bio /> renders. When bioText resolves an update will be made in state which will trigger a rerender in <Bio />, adding that text description to the page.

### Data fetching libraries

> !TODO React Query section

Reviewer: ```suggestion
We haven't even begun to scratch the surface of data fetching in the frontend. Keeping your frontend data most up-to-date with the server is a difficult task to accomplish. Managing "async" state becomes harder with every new feature. 

You've already tasted the complexity of data fetching in this lesson. Each request has to have a _minimum_ of three states to achieve an optimal user experience: `data`, `loading`, and `error`. There's a library called [TanStack Query](https://tanstack.com/query/latest/docs/react/overview) that helps you in data fetching and more. It keeps track of all the necessary states for you, and has built-in support for major features like caching.

Query is a great library to learn, however, we strongly suggest you stick to vanilla React data fetching for all the projects you build in this course. The lessons you will learn will be while doing so will be invaluable.

```

### Assignment

<div class="lesson-content__panel" markdown="1">

1. Read [Modern API data fetching methods](https://blog.logrocket.com/modern-api-data-fetching-methods-react/) for a brief overview of what was discussed in this lesson. 
2. Read [How to fetch data in React with performance in mind](https://www.developerway.com/posts/how-to-fetch-data-in-react) to learn more about efficiently handling fetch requests in React components.

</div>

### Knowledge check

This section contains questions for you to check your understanding of this lesson on your own. If you’re having trouble answering a question, click it and review the material it links to.

- <a class="knowledge-check-link" href="#using-fetch-in-react-components">How can you fetch data from an API in React?</a>
- <a class="knowledge-check-link" href="#handling-errors">Why should you manually throw errors in fetch requests?</a>
- <a class="knowledge-check-link" href="#managing-multiple-fetch-requests">How can you avoid waterfalling requests?</a>

## Additional resources

This section contains helpful links to related content. It isn’t required, so consider it supplemental.

> - It looks like this lesson doesn't have any additional resources yet. Help us expand this section by contributing to our curriculum.

Reviewer: ```suggestion
- [TanStack Query documentation](https://tanstack.com/query/latest/docs/react/overview)
```

### Introduction

Up to this point we have been using React to build client side applications with interactive user interfaces, but what if we want fetch data from the internet? In order to create full fledged web applications, we need some way to get data from external sources and dynamically displaying it on the client.

> In this lesson, we'll explore the ins and outs of fetching data in React, starting with the basics of making API calls, managing component state, and handling asynchronous operations using JavaScript's ~fetch~ function.

Reviewer: Since the learners would've already done data fetching in the memory card and shopping cart projects, it's good to mention it here.

```suggestion
In this lesson, we'll explore the ins and outs of fetching data in React, starting with the basics of making API calls, managing component state, and handling asynchronous operations using JavaScript's `fetch` function. You've already performed data fetching in earlier projects, so some stuff covered in the lesson will be familiar to you. Revision doesn't hurt!
```

### Lesson overview

This section contains a general overview of topics that you will learn in this lesson.

- Understand how to make fetch requests in React components.
- Using conditonals to render content depending on the state of the request (short circuiting).
- Catching and handling errors.
- Lifting requests up the component hierarchy.

### A basic fetch request

Before we dive into the specifics of fetching data in React, let's first revisit how we can use the fetch API to get data from a server.

> <!-- fetch is used to make requests to external APIs and returns a Promise that resolves to the response object representing the server's response. -->

Reviewer: I don't think we need this comment. Learners are already familiar with fetch and promises
```suggestion
```

~~~js
const image = document.querySelector("img");
fetch("https://jsonplaceholder.typicode.com/photos", {
  mode: "cors",
})
  .then((response) => response.json())
  .then((response) => {
    img.src = response[0].url;
  })
  .catch((error) => console.error(error));
~~~

> We're making a request to the JSONPlaceholder API to retrieve an image. The response is then converted to JSON format using the json() method, and then we set the image src to the url embedded in that response object. The catch method is used to catch any errors that may occur during the request.

Reviewer: Similarly here, the learners already understand the fetch, catch, and .json()
```suggestion
We're making a request to the JSONPlaceholder API to retrieve an image. You should be able to tell what the above code snippet does.
```

### Using fetch in React components

Now let's take a look at how we can incorporate fetch into a similar React component. One common use case is to fetch data from an API when a component mounts, so that the data can be displayed on screen.

Whenever a component needs to make a request as it renders, it's often best to wrap that fetch inside of an effect.

**https://codesandbox.io/s/top-simple-react-request-rd9psp?file=/src/styles.css**

~~~jsx
import { useEffect, useState } from "react";

const Image = () => {
  const [imageURL, setImageURL] = useState(null);

  useEffect(() => {
    fetch("https://jsonplaceholder.typicode.com/photos", { mode: "cors" })
      .then((response) => response.json())
      .then((response) => setImageURL(response[0].url))
      .catch((error) => console.error(error));
  }, []);

  return (
    imageURL && (
      <>
        <h1>An image</h1>
        <img src={imageURL} alt={"placeholder text"} />
      </>
    )
  );
};

export default Image;
~~~

> useState lets us bundle data within a component and rerender the component whenever state is updated, and useEffect allows us to perform side effects on those components. Since we need to synchronize this component's state (the image URL) with an unpredictable external system like the fetch API, we need an effect. In this case we only want to fetch the URL once when the component is first mounted so we pass an empty array as a dependency to useEffect.

Reviewer: Learners already know how `useState` and `useEffect` works. Wording suggestion: 
```suggestion
`useState` lets us add the `imageURL` state, whereas `useEffect` allows us to perform side effects. In this case, the side effect is fetching data from an external API. Since we only need to fetch once when component mounts, we use an empty dependency array.
```

> > Reminder: Whenever you are changing state, (or doing anything else that could cause a rerender) inside of useEffect, be sure to always pass a dependency. If you forget to do so useEffect will enter an infinite loop and rerender exponentially.
> 
> You might be wondering what's going on here
> 
> ~~~jsx
> return (
>   imageURL && (
>     <>
>       <h1>An image</h1>
>       <img src={imageURL alt={"placeholder text"}}>
>     </>
>   )
> )
> ~~~
> 
> This is a pattern called _short circuiting_ and it allows developers to conditionally render JSX without using if statements or ternarys. The above snippet can be read as, "if imageURL is truthy then continue to evaluate the expression". A truthy value and an object evaluates to the object so in this case, JSX is returned. If imageURL was falsy, the right hand side of the && operation wouldn't evaluate and the component would return false.

Reviewer: All of this the learner has learned in the previous lessons. I say let's remove it.
```suggestion
```

To fix this, we need to check _something_ in the return of the Image component. We'll call it "error".

~~~jsx
  return (
    <>
      {error && <p>A network error was encountered</p>}
      {imageURL && (
        <>
          <h1>An image</h1>
          <img src={imageURL} alt={"placeholder text"} />
        </>
      )}
    </>
  );
~~~

To set this error, we'll add it to the component's state.

~~~jsx
const [imageURL, setImageURL] = useState(null);
const [error, setError] = useState(null);
~~~

And finally, to give error a value when a request fails, we'll add a conditional to check the response status and set it where our console.error line was.

~~~jsx
useEffect(() => {
  fetch("https://jsonplaceholder.typicode.com/photos", { mode: "cors" })
    .then((response) => {
      if (response.status >= 400) {
        throw new Error("server error");
      }
      return response.json();
    })
    .then((response) => setImageURL(response[0].url))
    .catch((error) => setError(error));
}, []);
~~~

> > Note: A fetch request can return an http error without triggering the catch block. Always be sure to check your response and manually throw an error

Reviewer: This is something new I learned. Can you share a reference link so that I can read more on this?

Author: Certainly https://plainenglish.io/blog/handling-unsuccessful-calls-of-fetch-api-7eef551b74c8

It might be more accurate to say "A fetch request can return an http error **_status_** without triggering the catch block".  Catch will "catch" network errors and invalid URLs  but not > 400 status codes

Now when a bad URL is passed or the API returns an unexpected response, the page will relay that information to the user. 

**https://codesandbox.io/s/top-simple-react-request-with-error-fqjcrq?file=/src/Image.jsx**

### Managing multiple fetch requests

> In a full scale web app you're often going to be making more than one request and you need to be careful with how you organize them. A common issue new react developers face when their apps start making multiple requests is called a waterfall of requests. Let's look at an example.

Reviewer 2: ```suggestion
In a full scale web app you're often going to be making more than one request and you need to be careful with how you organize them. A common issue new React developers face when their apps start making multiple requests is called a "waterfall of requests". Let's look at an example.
```

React needs capital R, second wrapping in "" is merely a suggestion

Author: Totally agree!  The quotes don't look quite right.  I'll make that part italicized bold.

        .then((response) =>
          setBioText("I like long walks on the beach and JavaScript")
        )
        .catch((error) => console.error(error));
    }, delay + 2000); // here we add an extra 2 seconds of delay
  }, [delay]);

  return (
    (imageURL && (
      <div>
        <h3>Username</h3>
        <img src={imageURL} alt={"profile"} />
        <Bio bioText={bioText} />
      </div>
    )) || <h1>Loading...</h1>
  );
};

// Bio.jsx
const Bio = ({ bioText }) => {
  return (
    bioText && (
      <>
        <p>{bioText}</p>
      </>
    )
  );
};
~~~

Now we have both requests firing as soon as Profile renders. The request for imageURL resolves 2 seconds before the bioText request and our div containing <Bio /> renders. When bioText resolves an update will be made in state which will trigger a rerender in <Bio />, adding that text description to the page.

### Data fetching libraries

!TODO React Query section

### Assignment

<div class="lesson-content__panel" markdown="1">

1. Read [Modern API data fetching methods](https://blog.logrocket.com/modern-api-data-fetching-methods-react/) for a brief overview of what was discussed in this lesson. 

> 2. Read [How to fetch data in React with performace in mind](https://www.developerway.com/posts/how-to-fetch-data-in-react) to learn more about efficiently handling fetch requests in React components.

Reviewer 2: ```suggestion
2. Read [How to fetch data in React with performance in mind](https://www.developerway.com/posts/how-to-fetch-data-in-react) to learn more about efficiently handling fetch requests in React components.
```
